runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,517 @@
+                operations.forEach(Runnable::run);
+                operations.clear();
+                txnType = TxnType.writeAndReads;
+                break;
+            case 3: // commit called in a read-only transaction.
+                if (txnType == TxnType.readOnly) {
+                    log.trace("Ending a read-only transaction");
+                } else if (txnType == TxnType.writeAndReads) {
+                    log.trace("Ending a dirty-read transaction!");
+                } else { // Attempting to commit a corfu txn started outside this scope!
+                    log.error("commit() on a transaction started outside scope of this thread!");
+                    throw new IllegalStateException(
+                            "commit() invoked but transaction not started in this thread!");
+                }
+                break;
+            case 4:
+                log.trace("Dirty reads invoked. Applying {} in thread", operations.size());
+                tablesInTxn.add(tableBeingRead);
+                txBeginInternal();
+                operations.forEach(Runnable::run);
+                operations.clear();
+                txnType = TxnType.writeAndReads;
+                break;
+            case 5: // Normal path for a starting a transaction to do queries.
+                log.trace("Starting a transaction for reading data");
+                tablesInTxn.add(tableBeingRead);
+                txBeginInternal();
+                txnType = TxnType.readOnly;
+                break;
+            case 6:
+                log.trace("More dirty reads. Applying {}", operations.size());
+                tablesInTxn.add(tableBeingRead);
+                operations.forEach(Runnable::run);
+                txnType = TxnType.writeAndReads;
+                operations.clear();
+                break;
+            case 7:
+                tablesInTxn.add(tableBeingRead);
+                if (txnType != TxnType.writeAndReads) {
+                    log.trace("More reads in same transaction");
+                    txnType = TxnType.readOnly;
+                } else {
+                    log.trace("More reads in transaction that had writes before");
+                }
+                break;
+            default:
+                log.warn("Unsupported decision logic for txn begin");
+        }
+    }
+
+    /**
+     * Start the actual corfu transaction.
+     */
+    private void txBeginInternal() {
+        this.txnStartTime = System.nanoTime();
+        Transaction.TransactionBuilder transactionBuilder = this.objectsView
+                .TXBuild()
+                .type(TransactionType.WRITE_AFTER_WRITE);
+        if (isolationLevel.getTimestamp() != Token.UNINITIALIZED) {
+            transactionBuilder.snapshot(isolationLevel.getTimestamp());
+        }
+        transactionBuilder.build().begin();
+    }
+
+    /**
+     * Commit the transaction.
+     * For a transaction that only has write operations, this method will start and end
```

Comment:
I think this line is a bit confusing now that we have both reads and writes in the same transaction.

Reply:
modified comment.